### PR TITLE
#7224 Pull images parallel during deployment

### PR DESCRIPTION
**Pull request by AshwinMN24**

### CODE REVIEW

File: infrastructure/deployment/deploy.sh [hunk 1]
```diff
@@ -284,14 +312,13 @@ docker_stack_deploy() {
...
-      echo "Server failed to download $tag. Retrying..."
-      sleep 5
-    done
+    echo -e "$tag \n" >> docker_images.txt
   done
-
+  echo "-----"
+  cat docker_images.txt
+  echo "-----"
+  #configured_ssh "cd /opt/opencrvs && docker pull $tag"
```
Comment (rikukissa) [inline infrastructure/deployment/deploy.sh]: Let's clean up commits and debug messages from PRs

File: infrastructure/deployment/deploy.sh [hunk 2]
```diff
@@ -284,14 +312,13 @@ docker_stack_deploy() {
     fi
 
     echo "Downloading $tag"
-
-    until configured_ssh "cd /opt/opencrvs && docker pull $tag"
-    do
-      echo "Server failed to download $tag. Retrying..."
-      sleep 5
-    done
+    echo -e "$tag \n" >> docker_images.txt
```
Comment (rikukissa) [inline infrastructure/deployment/deploy.sh]: Is there a benefit for the extra step of first writing the `IMAGE_TAGS_TO_DOWNLOAD[@]` list to a file? Originally this loop's purpose was downloading the images, so we could keep it at that.

File: infrastructure/deployment/deploy.sh [hunk 3]
```diff
@@ -284,14 +312,13 @@ docker_stack_deploy() {
...
   done
-
+  echo "-----"
+  cat docker_images.txt
+  echo "-----"
+  #configured_ssh "cd /opt/opencrvs && docker pull $tag"
+  pull_images_from_file "docker_images.txt"
```
Comment (rikukissa) [inline infrastructure/deployment/deploy.sh]: Creating a file shouldn't be necessary for passing the tags forward, right? Use function arguments instead

File: infrastructure/deployment/deploy.sh [hunk 4]
```diff
@@ -269,6 +269,34 @@ split_and_join() {
    echo $SPLIT
 }
 
+pull_images_from_file() {
+    local file="$1"  # The file containing the Docker images
+    # Check if the file exists
+    if [[ ! -f "$file" ]]; then
+        echo "File not found: $file"
+        exit 1
+    fi
```
Comment (rikukissa) [inline infrastructure/deployment/deploy.sh]: If you generate code with ChatGPT (allowed, of course) please clean up formatting to match the rest of the file. Maybe there's a linter for shell scripts we could start using? The generated comments are usually pretty self-explanatory asa well so I tend to clean those away from the code just to reduce noise.

Besides that, it's good to properly read the generated code through. For instance in this code I highlighted, we are checking a file exists that we created a few steps backwards.

File: infrastructure/deployment/deploy.sh [hunk 5]
```diff
@@ -269,6 +269,34 @@ split_and_join() {
...
+    # Read each line (image name) from the file
+    while IFS= read -r image; do
+    {
+        image=$(echo "$image" | xargs)
+        # Skip empty lines and comments (lines starting with #)
+        if [[ -z "$image" || "$image" =~ ^# ]]; then
+            continue
+        fi
+        echo "Pulling $image..."
+        if docker pull "$image"; then
```
Comment (rikukissa) [inline infrastructure/deployment/deploy.sh]: On which machine does this image get pulled to? I looked at the images on the farajaland-dev machine right after deployment

```
riku@farajaland-dev:~$ docker images | grep fara
opencrvs/ocrvs-farajaland                       <none>                              5d511c748fe0   About an hour ago   1.54GB
opencrvs/ocrvs-farajaland                       3534c7a                             4e85e3bed115   About an hour ago   1.54GB
```

A moment later (docker swarm downloads images automatically later on in the process if images are not present). First one is the one that got deployed.

```
riku@farajaland-dev:~$ docker images | grep fara
opencrvs/ocrvs-farajaland                       <none>                              af7798ce0120   23 minutes ago      1.54GB
opencrvs/ocrvs-farajaland                       <none>                              5d511c748fe0   About an hour ago   1.54GB
opencrvs/ocrvs-farajaland                       3534c7a                             4e85e3bed115   About an hour ago   1.54GB
```

**Review by rikukissa — CHANGES_REQUESTED** [review]

File: infrastructure/deployment/deploy.sh (hunk 1 above)
Comment (AshwinMN24) [reply infrastructure/deployment/deploy.sh]: sure will do the changes.

File: infrastructure/deployment/deploy.sh (hunk 2 above)
Comment (AshwinMN24) [reply infrastructure/deployment/deploy.sh]: This file contains all the images list which are present on the machine

**Comment by AshwinMN24** [comment]
sure will do the changes as suggested.

**Comment by AshwinMN24** [comment]
reopening the PR

**Pull request closed without merging**